include/oneapi/tbb/collaborative_call_once.h

```diff
@@ -0,0 +1,222 @@
+public:
+    runner_lifetime_checker(once_runner& r) : m_runner(r) {
+        m_runner.increase_ref();
+    }
+    ~runner_lifetime_checker() {
+        m_runner.decrease_ref();
+    }
+};
+
+class collaborative_once_flag : no_copy {
+    enum state {uninitialized, done};
+    std::atomic<std::uintptr_t> my_state{ state::uninitialized };
+
+    template <typename Fn, typename... Args>
+    friend void collaborative_call_once(collaborative_once_flag& flag, Fn&& f, Args&&... args);
+
+    template <typename Fn>
+    void do_collaborative_call_once(Fn&& f) {
+        std::uintptr_t expected = my_state.load(std::memory_order_acquire);
+        once_runner local_runner;
+
+        while (expected != state::done) {
+            if (expected == state::uninitialized && my_state.compare_exchange_strong(expected, reinterpret_cast<std::uintptr_t>(&local_runner))) {
+                // winner thread
+                local_runner.make_runner();
+
+                auto local_expected = reinterpret_cast<std::uintptr_t>(&local_runner);
+
+                try_call([&] {
+                    local_runner.run_once(std::forward<Fn>(f));
+                }).on_exception([&] {
+                    while (!my_state.compare_exchange_strong(local_expected, state::uninitialized)) {
+                        local_expected = reinterpret_cast<std::uintptr_t>(&local_runner);
+                    }
+                });
+
+                while (!my_state.compare_exchange_strong(local_expected, state::done)) {
+                    local_expected = reinterpret_cast<std::uintptr_t>(&local_runner);
+                }
+                
+                break;
+            } else {
+                // moonlighting thread
+                do {
+                    auto max_value = expected | (max_nfs_size-1);
+                    expected = spin_wait_while_eq(my_state, max_value);
+                // "expected > state::done" prevents storing values, when state is uninitialized
+                } while (expected > state::done && !my_state.compare_exchange_strong(expected, expected + 1));
+
+                if (auto runner = reinterpret_cast<once_runner*>(maskoff_pointer(expected))) {
+
+                    runner_lifetime_checker hold_ref_count{*runner};
+                    my_state.fetch_sub(1);
+
+                    runner->wait_for_init();
+
+                    // The moonlighting threads are not expected to handle exceptions from user functor.
+                    // Therefore, no exception is expected from assist().
+                    runner->assist();
+                }
+            }
+        }
+    }
+};
+
+
+template <typename Fn, typename... Args>
+void collaborative_call_once(collaborative_once_flag& flag, Fn&& fn, Args&&... args) {
+    // Using stored_pack to suppress bug in GCC 4.8
+    // with parameter pack expansion in lambda
+    auto stored_pack = save_pack(std::forward<Args>(args)...);
+    auto func = [&] { call(std::forward<Fn>(fn), stored_pack); };
```

Comment:
I do not think that all compilers should suffer from `save_pack` because of GCC 4.8. Let's implement two versions: one for GCC 4.8 and another one is generic with variadic in lambdas.